### PR TITLE
Use functions for state script instead of aliases

### PR DESCRIPTION
**Pull request by shaunmlowry**
Using aliases in bash to add state tool scripts to the virtual
environments has a number of disadvantage.  One major drawback is that
these aliases are not available in subshells, or available to scripts.

That means doing things like:
```
$ myvar=$(mystatescript)
```
don't work.

Instead, create shell functions and export them with `export -f` so that
they can be used everywhere a regular command could be.

 https://www.pivotaltracker.com/story/show/178030472

### CODE REVIEW

File: assets/shells/bashrc.sh [hunk 1]
```diff
@@ -21,7 +21,10 @@ alias {{.ExecName}}='{{.ExecAlias}}'
 {{ end }}
 
 {{range $K, $CMD := .Scripts}}
-alias {{$K}}='{{$.ExecName}} run {{$CMD}}'
+function {{$K}} {
+    {{$.ExecName}} run {{$CMD}} "$@"
+}
+export -f {{$K}}
```
Comment (Naatan) [inline assets/shells/bashrc.sh:27]: I don't think we need this `export`, given this code will effectively be injected into your bashrc

**Review by Naatan — CHANGES_REQUESTED** [review]
Thanks! Do you know if this would also be relevant for zsh?

File: assets/shells/bashrc.sh (hunk 1 above)
Comment (shaunmlowry) [reply assets/shells/bashrc.sh:27]: .bashrc isn't sourced for things like `bash -c` or scripts that start with `#!/bin/bash` so the export makes the commands available in those situations

Comment (shaunmlowry) [reply assets/shells/bashrc.sh:27]: the environment inherited is just the environment variables which *nix operating systems understand at a lower level (see `execle()`'s `char *const envp[]` argument for example).  It doesn't include shell-specific items like aliases and functions.  `export -f` is a special variant in bash which tells it to copy functions from the current shell into spawned _bash_ shells specifically.   See `man bash` for details on how `export -f` works

**Review by Naatan — APPROVED** [review]

**Pull request merged**